delfin/drivers/dell_emc/vnx_block/component_handler.py

```diff
@@ -0,0 +1,477 @@
+            if objs:
+                for obj in objs:
+                    if obj.get('pool_name') is not None:
+                        capacity = float(obj.get("available_capacity_gbs", 0))
+                        obj_free += capacity
+                obj_model = {
+                    'obj_free': obj_free * units.Gi
+                }
+            return obj_model
+
+        except Exception as e:
+            err_msg = "Failed to get pool capacity from EmcVnxStor: %s" % (
+                six.text_type(e))
+            LOG.error(err_msg)
+            raise exception.InvalidResults(err_msg)
+
+    def handler_raid_group_capacity(self):
+        try:
+            # Get raid_group capacity
+            objs = self.navi_handler.get_raid_group()
+            obj_free = 0
+            obj_model = {}
+            if objs:
+                for obj in objs:
+                    if obj.get('raidgroup_id') is not None:
+                        capacity = float(
+                            obj.get("free_capacity_blocks,non-contiguous", 0))
+                        obj_free += capacity
+                obj_model = {
+                    'obj_free': obj_free * (units.Ki / 2)
+                }
+            return obj_model
+
+        except Exception as e:
+            err_msg = "Failed to get raid capacity from EmcVnxStor: %s" % (
+                six.text_type(e))
+            LOG.error(err_msg)
+            raise exception.InvalidResults(err_msg)
+
+    def handler_storage_free_capacity(self):
+        try:
+            obj_model = {}
+            # Get storage_free capacity
+            free_cap = 0
+            raw_cap = 0
+            pool_free = 0
+            raid_free = 0
+            try:
+                raw_cap = self.handler_disk_capacity()
+            except Exception:
+                LOG.error('Get disk capacity failed!')
+
+            try:
+                pool_map = self.handler_pool_capacity()
+                if pool_map:
+                    pool_free = pool_map.get('obj_free')
+            except Exception:
+                LOG.error('Get pool capacity failed!')
+
+            try:
+                raid_group_map = self.handler_raid_group_capacity()
+                if raid_group_map:
+                    raid_free = raid_group_map.get('obj_free')
+            except Exception:
+                LOG.error('Get raid group capacity failed!')
+
+            free_cap = pool_free + raid_free
+
+            obj_model = {
+                'raw_cap': raw_cap,
+                'free_cap': free_cap
+            }
+            return obj_model
+
+        except Exception as e:
+            err_msg = "Failed to get storage capacity from EmcVnxStor: %s" % (
+                six.text_type(e))
+            LOG.error(err_msg)
+            raise exception.InvalidResults(err_msg)
+
+    def handler_pool_lun_capacity(self):
+        try:
+            # Get pool lun capacity
+            objs = self.navi_handler.get_pool_lun()
```

Comment:
Do we need to parse through the LUNs to get pool leval capacity? cant we get it directly?